File: android-native/app/src/main/java/com/example/kpkn/screens/workout/WorkoutUiCommon.kt
package com.example.kpkn.screens.workout

import android.annotation.SuppressLint
import android.content.Context
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import com.example.kpkn.data.models.HomologatedPerformanceResult
import kotlin.math.abs

internal fun Double.toTrimmedNumberString(): String {
    val rounded = ((this * 10).toInt()) / 10.0
    return if (rounded == rounded.toInt().toDouble()) {
        rounded.toInt().toString()
    } else {
        rounded.toString()
    }
}

internal fun formatSignedDelta(value: Double, suffix: String = ""): String {
    val absValue = abs(value)
    val base = absValue.toTrimmedNumberString()
    val unit = if (suffix.isBlank()) "" else suffix
    return when {
        value > 0.0 -> "+$base$unit"
        value < 0.0 -> "-$base$unit"
        else -> "0$unit"
    }
}

@SuppressLint("MissingPermission")
internal fun triggerPRCelebrationHaptic(context: Context) {
    val waveform = longArrayOf(0L, 100L, 150L, 100L, 300L)
    runCatching {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val manager = context.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager
            val vibrator = manager.defaultVibrator
            vibrator.vibrate(VibrationEffect.createWaveform(waveform, -1))
        } else {
            @Suppress("DEPRECATION")
            val vibrator = context.getSystemService(Context.VIBRATOR_SERVICE) as? Vibrator
            vibrator?.let {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    it.vibrate(VibrationEffect.createWaveform(waveform, -1))
                } else {
                    @Suppress("DEPRECATION")
                    it.vibrate(waveform, -1)
                }
            }
        }
    }
}

internal fun formatTime(seconds: Int): String {
    val m = seconds / 60
    val s = seconds % 60
    return if (m > 0) "${m}:${s.toString().padStart(2, '0')}" else "${s}s"
}

internal fun formatElapsed(seconds: Int): String {
    val hours = seconds / 3600
    val minutes = (seconds % 3600) / 60
    val secs = seconds % 60
    return if (hours > 0) {
        "%d:%02d:%02d".format(hours, minutes, secs)
    } else {
        "%02d:%02d".format(minutes, secs)
    }
}

internal fun buildWorkoutAchievementMessage(
    homologated: HomologatedPerformanceResult?,
    showPRsInWorkout: Boolean,
): String? {
    if (homologated == null) return null
    if (!showPRsInWorkout) return homologated.suggestionReason?.takeIf { it.isNotBlank() }

    val metric = homologated.metricValue.toTrimmedNumberString()
    val label = homologated.metricType.ifBlank { "Rendimiento" }

    return when {
        homologated.isGlobalPr -> "PR global · $label $metric"
        homologated.isContextPr -> "PR contextual · $label $metric"
        !homologated.suggestionReason.isNullOrBlank() -> homologated.suggestionReason
        else -> null
    }
}

File: ios-native/KPKNFit/KPKNFit/Presentation/Screens/Workout/WorkoutUiCommon.swift
import Foundation
#if canImport(UIKit)
import UIKit
#endif

// MARK: - Extension to format Doubles nicely
extension Double {
    func toTrimmedNumberString() -> String {
        let rounded = Double(Int(self * 10)) / 10.0
        if rounded == Double(Int(rounded)) {
            return String(Int(rounded))
        } else {
            return String(rounded)
        }
    }
}

// MARK: - Signed Delta Formatting
func formatSignedDelta(value: Double, suffix: String = "") -> String {
    let absValue = abs(value)
    let base = absValue.toTrimmedNumberString()
    let unit = suffix.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? "" : suffix
    if value > 0.0 {
        return "+\(base)\(unit)"
    } else if value < 0.0 {
        return "-\(base)\(unit)"
    } else {
        return "0\(unit)"
    }
}

// MARK: - PR Haptic triggers
func triggerPRCelebrationHaptic(context: Any? = nil) {
    #if os(iOS)
    let generator = UINotificationFeedbackGenerator()
    generator.prepare()
    generator.notificationOccurred(.success)
    #endif
}

// MARK: - Time format helper (m:ss or ss)
func formatTime(seconds: Int) -> String {
    let m = seconds / 60
    let s = seconds % 60
    if m > 0 {
        return "\(m):\(String(format: "%02d", s))"
    } else {
        return "\(s)s"
    }
}

// MARK: - Elapsed Timer helper
func formatElapsed(seconds: Int) -> String {
    let hours = seconds / 3600
    let minutes = (seconds % 3600) / 60
    let secs = seconds % 60
    if hours > 0 {
        return String(format: "%d:%02d:%02d", hours, minutes, secs)
    } else {
        return String(format: "%02d:%02d", minutes, secs)
    }
}

// MARK: - Achievement Message formatter
func buildWorkoutAchievementMessage(
    homologated: HomologatedPerformanceResult?,
    showPRsInWorkout: Bool
) -> String? {
    guard let homologated = homologated else { return nil }
    if !showPRsInWorkout {
        if let reason = homologated.suggestionReason, !reason.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return reason
        }
        return nil
    }

    let metric = homologated.metricValue.toTrimmedNumberString()
    let label = homologated.metricType.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? "Rendimiento" : homologated.metricType

    if homologated.isGlobalPr {
        return "PR global · \(label) \(metric)"
    } else if homologated.isContextPr {
        return "PR contextual · \(label) \(metric)"
    } else if let reason = homologated.suggestionReason, !reason.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
        return reason
    } else {
        return nil
    }
}
